src/cirrus/builtins/functions/process/lambda_function.py

```diff
@@ -22,8 +24,15 @@ def lambda_handler(event, context):
 
         try:
             payload = utils.extract_record(message)
-        except Exception:
-            logger.exception("Failed to extract record: %s", json.dumps(message))
+        except Exception as exc:
+            logger.exception("Failed to extract record: %s", message)
+            with WorkflowEventManager.handler(logger=logger) as wfem:
+                wfem.announce(
+                    WFEventType.RECORD_EXTRACT_FAILED,
+                    payload=message,
+                    payload_id="unknown",
+                    extra_message={"exception": str(exc)},
+                )
```

Comment:
This appears to create a new workflow event manager just for this exception; if there are multiple exceptions you end up sending each event individually, right? Same with the other exception handler? You also have a wfem in process payload or something that is handling all other events?

Perhaps this is premature optimization, so feel free to disregard if you think so, but it _feels_ like someone makes a mistake (sending in a bunch of malformed payloads) and this becomes an problematic bottleneck. Could you instead use one wfem for all of these places you need to fire off events by creating one instance for this function, and passing it down into the `ProcessPayloads(payloads).process()` call?

Doing so seems like it would be annoying because you'd have to indent everything in this function to have it within the context of the wfem. I'd instead suggest a decorator pattern--you could add a method to the `WorkflowEventManager` class that implements a decorator that injects a keyword-only arg `wfem` into the wrapped function, and then you could use that decorator to inject `wfem` into this `lambda_handler` function.

Just a thought. We can discuss if it seems worth pursuing.

Reply:
Pursued, paired, and implemented!